### DISCUSSION TITLE
How to correctly use $refs?

### DISCUSSION THREAD

**Discussion by petasis** [alpinejs/alpine#3715] [1. Help]
Hi all,

I have a set of tables in my page, to which I want to deliver an event, if something is changed in a store.

However, I don't know how to pass $refs when I create the store:

```
<div class="container-fluid py-4" x-data="dashboard()" x-init="init()">
  <div x-ref="dashboard" />
</div>

<script>
function dashboard() {
    return { init: function() {
        Alpine.store('dashboardInteractiveFiltering', {
            // selected: $persist(new Object()).using(sessionStorage),
            selected: new Object(),

            addPK(vast_class, pk, checked=true) {
                if (this.selected[vast_class] == undefined) {
                    this.selected[vast_class] = new Set();
                }
                if (checked) {
                    this.selected[vast_class].add(pk);
                } else {
                    this.selected[vast_class].delete(pk);
                }
                console.error(vast_class, this.selected[vast_class])
                this.$refs.dashboard.dispatchEvent(
                    new Event('selection-changed', { bubbles: true, detail: {} }));
            }
        });
    }};
};
</script>
```

The error I get is that `this.$refs` is undefined.
this$refs exists in the init anonymous function. But not inside Alpine.store function.

Will $refs be able to also list components that are dashboard's children?

**Comment by petasis** [comment]
It seems that if I add `$refs = this.$refs;` just before Alpine.store(), and change `this.$refs` to `$refs` inside store(), seems to work.

Is this safe to do?

**Reply by SimoTod** [reply]
Refs is just a querySelector kind of thing, it's hardly a good solution for any use case. 

You probably want to rethink your design. 

If you want to use events, don't enumerate all the components that are listening. Just dispatch the event on the current element, let it bubble up to window and then all the listening components can just use the window modifier to observe the event.
But also, if you use a store, you don't need events at all, everything will react automatically to any store change without the need of notifying the components.

**Comment by ekwoka** [comment]
$refs are element bound, so they aren't available in stores.

The biggest question becomes: Why do you want to dispatch events on the DOM like this? Can the rows not just respond to the store value themselves?

**Reply by petasis** [reply]
Well, I have given up on $refs, they don't work.

Now, my problem (on which I have spend hours searching without a result), is how to access the $store from plain javascript that runs on the global scope.

How can be done?

I have this object in the `$store.dashboardInteractiveFiltering.selected`. How can I pass it to HTMx?

`hx-vals='javascript:{data: <what to put here?>}'` 

this show the window object when executed by HTMx.

**Reply by ekwoka** [reply]
Well, $refs work, just maybe not to do the thing you are trying to do.

But `Alpine.store` is the way.

It's a polymorphic function, so passing it a string and object creates a store, passing just a string, returns the store of that name.

Not a totally uncommon api shape, but also not the most intuitive.

**Comment by petasis** [comment]
It seems that this works:

```
hx-vals="javascript:{selectionData: Alpine.store('dashboardInteractiveFiltering').selected}"
```

**Reply by SimoTod** [reply]
Yep, Alpine.store 👍
